### ISSUE TITLE
including startup time makes the results misleading

### ISSUE THREAD

**Issue by stevengj** [niklas-heer/speed-comparison#130]
In Julia, and perhaps in some of the other languages, you are mostly measuring the one-time startup cost, rather than the cost of actually doing the mathematical calculation.   So the results are misleading as a guide for speed of realistic number-crunching tasks, since compute-heavy problems inevitably run for long enough that the startup time is negligible.

In Julia's case, this startup cost is especially heavy because it includes a one-time cost of compiling the function you are benchmarking.   It would be like including the time for the C compiler in the C benchmark.

Recommendation: do the timing *within* each language's script (all languages offer high-resolution timers), and print out the elapsed compute time for calling `f(rounds)` *only*.     In Julia's case, call the function once (e.g. with `rounds=1`) before beginning the timing, to trigger its compilation.

**Comment by nsajko** [comment]
> using StaticCompiler.jl (although this is a typical way to run Julia)

There's a "not" missing there.

**Comment by stevengj** [comment]
fixed

**Comment by Moelf** [comment]
https://github.com/niklas-heer/speed-comparison/issues/22

**Comment by 0xF48** [comment]
durrrr....the timer has to be before and after the function call

**Comment by yonnig** [comment]
I completely agree with [stevengj's post](https://github.com/niklas-heer/speed-comparison/issues/130#issue-2539646673), and one might wonder if this question also applies to Python and other languages of the "same family".

Contrary to what is written in [a post of the linked thread](https://github.com/niklas-heer/speed-comparison/issues/22#issuecomment-1279822367), personal opinions have nothing to do here. The goal should be to present the results as comprehensively and transparently as possible, as in any scientific approach. This is all the more regrettable since these benchmarks appear very prominently in search engine results, thus giving them a certain legitimacy.

So, for languages ​​where compilation time is required for a first run, why not present two results: the first for a cold run, and the second for a hot run?

**Comment by ShalokShalom** [comment]
This seems unmaintained for years, so if nobody is going to fork and run the results on their own, I doubt this will be resolved.

**Comment by PallHaraldsson** [comment]
@ShalokShalom, not sure "unmaintained" (repo), I see activity today from @niklas-heer and actually last about a yer ago).

**Comment by niklas-heer** [comment]
Thanks for raising this issue and the thoughtful discussion! You're absolutely right that startup/JIT compilation overhead was disproportionately affecting the results, especially for languages like Julia.

I recently overhauled the CI pipeline with matrix builds and aggressive caching (both for the benchmark tool and individual language results). This means benchmarks only re-run when source code actually changes, making longer run times much more feasible.

Regarding the suggestions:

**1. In-process timing with language-native timers** - This would require modifying all 46 language implementations, each with different timing APIs. While more accurate, it's a significant maintenance burden.

**2. Pre-warming compiled code** - I already run warmup iterations, but since each benchmark spawns a fresh process, languages like Julia still incur JIT overhead every time.

**3. Report minimum time** - Already implemented! I report min, median, max, and mean.

**4. Increase iterations** - This is what I went with. I've bumped iterations from 100 million to 1 billion (10x). This means:
- Fast languages (C, Zig, Rust): ~3-30 seconds
- Medium languages (Java, PyPy, Node): ~10-20 seconds  
- Slow languages (CPython, Ruby, Lua): 2-7 minutes

With these run times, startup overhead becomes negligible compared to actual computation time.

The change is live now: https://github.com/niklas-heer/speed-comparison/commit/9ef12e6

**Comment by niklas-heer** [comment]
The new results with 1 billion iterations are now live:

📊 **Latest results:** https://github.com/niklas-heer/speed-comparison/blob/master/docs/history/latest/combined_results.png

![Results](https://raw.githubusercontent.com/niklas-heer/speed-comparison/master/docs/history/latest/combined_results.png)

**Comment by ShalokShalom** [comment]
I guess Odin and Roc fit into this comparison, as they are one of the few exotic languages that can compete on the top.

**Comment by niklas-heer** [comment]
Good news @ShalokShalom - Odin has just been added! It's included in the upcoming benchmark run along with several other new languages (Dart, Deno, Erlang, Gleam, Janet, Racket, V, and WASM).

Roc is an interesting suggestion too - it's still in early development but could be worth adding once it stabilizes further. Thanks for the recommendation!

**Comment by niklas-heer** [comment]
The benchmark results are now in! Odin performed excellently at **1,402ms** - right in the same tier as Go, Pascal, Scala, and OCaml. Great call on that suggestion!

V also came in at the same tier (1,403ms), and Dart AOT is competitive with Swift at 1,407ms.

**Comment by GoNZooo** [comment]
> The benchmark results are now in! Odin performed excellently at **1,402ms** - right in the same tier as Go, Pascal, Scala, and OCaml. Great call on that suggestion!
> 
> V also came in at the same tier (1,403ms), and Dart AOT is competitive with Swift at 1,407ms.

Honestly, having Odin in the benchmark with that result isn't doing it any favors, since it performs the same as C, Zig and C++. I think it's best to keep languages where a good, representative implementation cannot currently be written out of the benchmark.

**Comment by ShalokShalom** [comment]
> > The benchmark results are now in! Odin performed excellently at **1,402ms** - right in the same tier as Go, Pascal, Scala, and OCaml. Great call on that suggestion!
> > V also came in at the same tier (1,403ms), and Dart AOT is competitive with Swift at 1,407ms.
> 
> Honestly, having Odin in the benchmark with that result isn't doing it any favors, since it performs the same as C, Zig and C++. I think it's best to keep languages where a good, representative implementation cannot currently be written out of the benchmark.

You see Odin misrepresented, since it is unable to implement that specific benchmark, and is just as performant as Zig in other benchmarks?

**Comment by GoNZooo** [comment]
> > > The benchmark results are now in! Odin performed excellently at **1,402ms** - right in the same tier as Go, Pascal, Scala, and OCaml. Great call on that suggestion!
> > > V also came in at the same tier (1,403ms), and Dart AOT is competitive with Swift at 1,407ms.
> > 
> > 
> > Honestly, having Odin in the benchmark with that result isn't doing it any favors, since it performs the same as C, Zig and C++. I think it's best to keep languages where a good, representative implementation cannot currently be written out of the benchmark.
> 
> You see Odin misrepresented, since it is unable to implement that specific benchmark, and is just as performant as Zig in other benchmarks?

I view this as a misrepresentation because I write 3D engines in Odin and the implication that Odin somehow is not as performant as C, Zig, et al, is plain false. If we could somehow get better performance by using Zig we would, but having used that for several years I can tell you that it's never going to be the case.

To clarify, so there can be no misunderstanding: If someone is unable to write a representative example for this collection of benchmarks they should refrain from having the language in the lineup. It's not about the language at all; I know from experience it has the same exact performance characteristics as Zig, C, etc.; it's about the person that writes the implementation. Other people don't know this, and this repository is going to end up being misleading.

**Comment by ShalokShalom** [comment]
> I view this as a misrepresentation because I write 3D engines in Odin and the implication that Odin somehow is not as performant as C, Zig, et al, is plain false. If we could somehow get better performance by using Zig we would, but having used that for several years I can tell you that it's never going to be the case.
> 
> To clarify, so there can be no misunderstanding: If someone is unable to write a representative example for this collection of benchmarks they should refrain from having the language in the lineup. It's not about the language at all; I know from experience it has the same exact performance characteristics as Zig, C, etc.; it's about the person that writes the implementation. Other people don't know this, and this repository is going to end up being misleading.

I understand. I could post this in the Odin community channels; maybe someone who is already known to the details of Odin could contribute to the repo. 

Micro benchmarks will always be micro benchmarks.
